lib/galaxy/tools/parameters/basic.py

```diff
@@ -921,9 +925,14 @@ def from_json(self, value, trans, other_values={}, require_legal_value=True):
                 raise ParameterValueError("multiple values provided but parameter is not expecting multiple values", self.name, is_dynamic=self.is_dynamic)
             rval = []
             for v in value:
-                if v not in legal_values:
-                    raise ParameterValueError("an invalid option ({!r}) was selected (valid options: {})".format(v, ",".join(legal_values)), self.name, v, is_dynamic=self.is_dynamic)
-                rval.append(v)
+                if v in legal_values:
+                    rval.append(v)
+            if len(rval) == 0:
+                for v in value:
+                    if v in fallback_values:
+                        rval.append(fallback_values[v])
+            if len(rval) == 0:
+                raise ParameterValueError("an invalid option ({!r}) was selected (valid options: {})".format(v, ",".join(legal_values)), self.name, v, is_dynamic=self.is_dynamic)
```

Comment:
```suggestion
                if v in legal_values:
                    rval.append(v)
                elif v in fallback_values:
                    rval.append(fallback_values[v])
                else:
                    raise ParameterValueError("an invalid option ({!r}) was selected (valid options: {})".format(v, ",".join(legal_values)), self.name, v, is_dynamic=self.is_dynamic)
```
All values need to be valid, while in your implementation just one value has to be valid.

Reply:
The intention was that all values should be taken from the same column. 

Maybe we can change this to something along the following lines:

```
if set(value).issubset(set(legal_values)):
    return value
elif set(value).issubset(set(fallback_values.keys())):
    return fallback_values.values()
else:
    raise ParameterValueError("invalid options ({!r}) were selected (valid options: {})".format(",".join(set(value)-set(legal_values)), ",".join(legal_values)), self.name, v, is_dynamic=self.is_dynamic)
```

Reply:
Seems that this is the last open point?

Reply:
I have just translated the existing code to also take into account `fallback_values`. I don't follow the column argument. Your alternative suggestion fails for string values:
```
In [1]: set('abcedef')
Out[1]: {'a', 'b', 'c', 'd', 'e', 'f'}
```

Reply:
I think `value` is always a list (or some other iterable) otherwise the for loop `for v in value:` in the original code would have failed, or? So with the small fix to use `.keys()` also for legal_values it should work:

```
if set(value).issubset(set(legal_values.keys())):
    return value
elif set(value).issubset(set(fallback_values.keys())):
    return fallback_values.values()
else:
    raise ParameterValueError("invalid options ({!r}) were selected (valid options: {})".format(",".join(set(value)-set(legal_values)), ",".join(legal_values)), self.name, v, is_dynamic=self.is_dynamic)
```

So this should test if all values are legal values or if all values are legal names. 

We just need to decide if this is the semantics that we want or your suggestion that all values need to be either a legal value or a legal name.

Reply:
You're right. I have no preference, but `legal_values` is a list, so should this be
```python
if set(value).issubset(set(legal_values)):
    return value
elif set(value).issubset(set(fallback_values.keys())):
    return fallback_values.values()
else:
    raise ParameterValueError("invalid options ({!r}) were selected (valid options: {})".format(",".join(set(value)-set(legal_values)), ",".join(legal_values)), self.name, v, is_dynamic=self.is_dynamic)
```
?

Reply:
Done. Note that `get_legal_values` returns a `set`. So it should be fine as is.